README.md
# BGZFStreams

[![Build Status](https://travis-ci.org/BioJulia/BGZFStreams.jl.svg?branch=master)](https://travis-ci.org/BioJulia/BGZFStreams.jl)
[![codecov.io](http://codecov.io/github/BioJulia/BGZFStreams.jl/coverage.svg?branch=master)](http://codecov.io/github/BioJulia/BGZFStreams.jl?branch=master)

BGZF is a compression format that supports random access using *virtual file offsets*.

See the SAM/BAM file format specs for the details of BGZF: <https://samtools.github.io/hts-specs/SAMv1.pdf>.


## Synopsis

```julia
using BGZFStreams

# The first argument is a filename or an IO object (e.g. IOStream).
stream = BGZFStream("data.bgz")

# BGZFStream is a subtype of IO and works like a usual IO object.
while !eof(stream)
    byte = read(stream, UInt8)
    # do something...
end

# BGZFStream is also seekable with a VirtualOffset.
seek(stream, VirtualOffset(0, 2))

# The current virtual file offset is available.
virtualoffset(stream)

close(stream)
```


## Usage

The BGZFStreams.jl package exports three types and a function to the package user:

* Types:
    * `BGZFStream`: an `IO` stream of the BGZF file format
    * `VirtualOffset`: data offset in a BGZF file
    * `BGZFDataError`: an error type thrown when reading a malformed byte stream
* Function:
    * `virtualoffset(stream)`: returns the current virtual file offset of `stream`

The `BGZFStream` type wraps an underlying `IO` object and transparently inflate
(for reading) or deflate (for writing) data. Since it is a subtype of `IO`, an
instance of it behaves like other `IO` objects, but the `seek` method takes a
virtual offset instead of a normal file offset as its second argument.

The `VirtualOffset` type represents a 64-bit virtual file offset as described in
the specification of the SAM file format. That is, the most significant 48-bit
integer of a virtual offset is a byte offset to the BGZF file to the beginning
position of a BGZF block and the least significant 16-bit integer is a byte
offset to the uncompressed byte(s).

The `BGZFDataError` type is a subtype of `Exception` and used to throw an
exception when invalid data are read.

The `virtualoffset(stream::BGZFStream)` returns the current virtual file offset.
More specifically, it returns the virtual offset of the next reading byte while
reading and the next writing byte while writing.


## Parallel Processing

A major bottleneck of processing a BGZF file is the inflation and deflation
process. The throughput of reading data is ~100 MiB/s, which is quite slower
than that of raw reading from a file. In order to alleviate this problem, this
package supports parallelized inflation when reading compressed data. This
requires the support of multi-threading introduced in Julia 0.5. The
`JULIA_NUM_THREADS` environmental variable sets the number of threads used for
processing.

    bash-3.2$ JULIA_NUM_THREADS=2 julia -q
    julia> using Base.Threads

    julia> nthreads()
    2


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2018-06-22
### Added
- Support for do-block syntax with BGZFStreams. Thanks @cth :smile:.

### Changed
- Up to date project files.
- Fix support for STDIN pipes. Thanks @dcjones :smile:.

### Dependencies
- :exclamation: Dropped support for julia v0.5.


## [0.1.4] - 2017-06-27
### Added
- Read and write methods for VirtualOffset

## [0.1.3] - 2017-03-02
### Changed
- :exclamation: Dropped support for julia v0.5 pre-releases.


## [0.1.2] - 2016-09-13
### Added
- Package now gets precompiled.


## [0.1.1] - 2016-09-10
### Changed
- Performance improvements using @inbounds annotated loops.


## [0.1.0] - 2016-09-08
### Added
- Support for reading data from a pipe.
- testing support for julia v0.5.


## [0.0.1] - 2016-07-30
- The initial release.


[Unreleased]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.1.4...v0.2.0
[0.1.4]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/BioJulia/BGZFStreams.jl/compare/v0.0.1...v0.1.0
[0.0.1]: https://github.com/BioJulia/BGZFStreams.jl/tree/v0.0.1

test/runtests.jl
using BGZFStreams
using Base.Test
using Compat

@testset "VirtualOffset" begin
    voff = VirtualOffset(0, 0)
    @test voff == VirtualOffset(0, 0)
    @test voff != VirtualOffset(0, 1)
    @test voff <= VirtualOffset(0, 1)
    @test voff != VirtualOffset(1, 0)
    @test voff <= VirtualOffset(1, 0)
    @test voff[1] == 0
    @test voff[2] == 0
    @test string(voff) == "BGZFStreams.VirtualOffset(0, 0)"

    voff += 1
    @test voff[1] == 0
    @test voff[2] == 1

    voff = VirtualOffset(1234, 555)
    @test voff[1] == 1234
    @test voff[2] == 555
    @test string(voff) == "BGZFStreams.VirtualOffset(1234, 555)"

    voff = VirtualOffset(1234, 555)
    buf = IOBuffer()
    @test write(buf, voff) == sizeof(UInt64)
    seekstart(buf)
    @test read(buf, VirtualOffset) === voff
    @test eof(buf)

    @test_throws ArgumentError VirtualOffset(1 << 48, 0)
    @test_throws ArgumentError VirtualOffset(0, 1 << 16)
end

@testset "BGZFStream" begin
    filename = joinpath(dirname(@__FILE__), "bar.bgz")
    stream = BGZFStream(filename, "r")
    @test read(stream, UInt8) === UInt8('b')
    @test read(stream, UInt8) === UInt8('a')
    @test read(stream, UInt8) === UInt8('r')
    @test eof(stream)
    @test flush(stream) === nothing
    @test close(stream) === nothing
    @test string(stream) == "BGZFStreams.BGZFStream{IOStream}(<mode=read>)"
    @test_throws ArgumentError read(stream, UInt8)

    stream = BGZFStream(filename, "r")
    data = zeros(Int8, 3)
    unsafe_read(stream, pointer(data), 3)
    @test data == UInt8['b', 'a', 'r']
    close(stream)
    @test_throws ArgumentError unsafe_read(stream, pointer(data), 3)

    open(BGZFStream, filename, "r") do stream
        @test virtualoffset(stream) === VirtualOffset(0, 0)
        read(stream, UInt8)
        read(stream, UInt8)
        @test virtualoffset(stream) === VirtualOffset(0, 2)
        seek(stream, VirtualOffset(0, 1))
        @test read(stream, UInt8) === UInt8('a')
        @test read(stream, UInt8) === UInt8('r')
        seekstart(stream)
        @test read(stream, UInt8) === UInt8('b')
        @test read(stream, UInt8) === UInt8('a')
       @test read(stream, UInt8) === UInt8('r')
    end

    # Empty data.
    empty_block = copy(BGZFStreams.EOF_BLOCK)
    stream = BGZFStream(IOBuffer(empty_block))
    @test_throws EOFError read(stream, UInt8)
    @test_throws EOFError read(stream, UInt8)
    stream = BGZFStream(IOBuffer(vcat(empty_block, empty_block)))
    @test_throws EOFError read(stream, UInt8)
    @test_throws EOFError read(stream, UInt8)
    stream = BGZFStream(IOBuffer(empty_block))
    @test isempty(read(stream))
    @test isempty(read(stream))
    stream = BGZFStream(IOBuffer(vcat(empty_block, empty_block)))
    @test isempty(read(stream))
    @test isempty(read(stream))

    # invalid data
    stream = BGZFStream(IOBuffer(vcat([0x1f, 0x00, 0x08, 0x04], rand(UInt8, 6))))
    @test_throws BGZFDataError read(stream)  # invalid gzip flag
    stream = BGZFStream(IOBuffer(vcat([0x1f, 0x8b, 0x00, 0x04], rand(UInt8, 6))))
    @test_throws BGZFDataError read(stream)  # invalid compression method
    stream = BGZFStream(IOBuffer(vcat([0x1f, 0x8b, 0x08, 0x01], rand(UInt8, 6))))
    @test_throws BGZFDataError read(stream)  # invalid flag

    filename = tempname()
    try
        stream = BGZFStream(filename, "w")
        @test virtualoffset(stream) == VirtualOffset(0, 0)
        @test write(stream, 'b') === 1
        @test write(stream, 'a') === 1
        @test write(stream, 'r') === 1
        @test virtualoffset(stream) == VirtualOffset(0, 3)
        @test write(stream, "baz") === 3
        @test eof(stream)
        @test flush(stream) === nothing
        @test close(stream) === nothing
        @test string(stream) == "BGZFStreams.BGZFStream{IOStream}(<mode=write>)"
    catch
        rethrow()
    finally
        if isfile(filename)
            rm(filename)
        end
    end

    # Pipe I/O
    filename = joinpath(dirname(@__FILE__), "bar.bgz")
    @test open(p -> read(BGZFStream(p)), `cat $(filename)`) == UInt8['b', 'a', 'r']
    @test_throws ArgumentError open(p -> virtualoffset(BGZFStream(p)), `cat $(filename)`)
    @test_throws ArgumentError open(p -> seek(BGZFStream(p), VirtualOffset(0, 0)), `cat $(filename)`)
    # TODO: test writing to a pipe?

    buffer = IOBuffer()
    stream = BGZFStream(buffer, "w")
    write(stream, "foo")
    close(stream)
    @test !isopen(buffer)
    @test_throws ArgumentError write(stream, 0x01)

    # Round trip
    for n in [0, 1, 2, 5, 10, 50, 100, 10_000, 100_000, 1_000_000]
        data = rand(0x00:0xf0, n)

        # bulk read/write
        buffer = IOBuffer()
        stream = BGZFStream(buffer, "w")
        # HACK: do not close the buffer after the stream is closed
        stream.onclose = io -> nothing
        write(stream, data)
        close(stream)
        seekstart(buffer)
        stream = BGZFStream(buffer)
        @test data == read(stream)
        close(stream)

        # read/write byte by byte
        buffer = IOBuffer()
        stream = BGZFStream(buffer, "w")
        # HACK: do not close the buffer after the stream is closed
        stream.onclose = io -> nothing
        for x in data
            write(stream, x)
        end
        close(stream)
        seekstart(buffer)
        stream = BGZFStream(buffer)
        data′ = UInt8[]
        while !eof(stream)
            push!(data′, read(stream, UInt8))
        end
        @test data == data′
        close(stream)
    end
end


src/tabix.jl
# Tabix
# =====
#
# Generic index for tab-delimited files.
#
# Li, Heng. "Tabix: fast retrieval of sequence features from generic
# TAB-delimited files." Bioinformatics 27.5 (2011): 718-719.
# Specification: http://samtools.github.io/hts-specs/tabix.pdf

# Binning index
typealias BinIndex Dict{UInt32,Vector{Chunk}}

# Linear index
typealias LinearIndex Vector{VirtualOffset}

type PseudoBin
    unmapped::Chunk
    n_mapped::UInt64
    n_unmapped::UInt64
end

type Tabix
    format::Int32
    columns::NTuple{3,Int}
    meta::Char
    skip::Int
    names::Vector{String}
    indexes::Vector{Tuple{BinIndex,LinearIndex,PseudoBin}}
    n_no_coor::Nullable{UInt64}
end

function Base.show(io::IO, index::Tabix)
    println(io, summary(index), ":")
    println(io, "  format: ", format2str(index.format))
    println(io, "  columns: ", index.columns)
    println(io, "  meta char: '", index.meta, "'")
    println(io, "  skip lines: ", index.skip)
      print(io, "  names: ", index.names)
end

"""
    Tabix(filename::AbstractString)
    Tabix(input::IO)
    read(filename::AbstractString, ::Type{Tabix})
    read(input::IO, ::Type{Tabix})

Load a Tabix index from a file.
"""
function Tabix(filename::AbstractString)
    return read(filename, Tabix)
end

function Tabix(input::IO)
    return read(input, Tabix)
end

function Base.read(filename::AbstractString, ::Type{Tabix})
    return open(input -> read(input, Tabix), filename)
end

function Base.read(input_::IO, ::Type{Tabix})
    input = BGZFStream(input_)

    # magic
    T = read(input, UInt8)
    B = read(input, UInt8)
    I = read(input, UInt8)
    x = read(input, UInt8)
    if T != UInt8('T') || B != UInt8('B') || I != UInt8('I') || x != 0x01
        error("invalid tabix magic bytes")
    end

    n_refs = read(input, Int32)
    format = read(input, Int32)
    col_seq = read(input, Int32)
    col_beg = read(input, Int32)
    col_end = read(input, Int32)
    meta = read(input, Int32)
    skip = read(input, Int32)

    l_nm = read(input, Int32)
    data = read(input, UInt8, l_nm)
    names = split(String(data), '\0', keep=false)

    indexes = Tuple{BinIndex,LinearIndex,PseudoBin}[]
    for n in 1:n_refs
        binindex = BinIndex()
        n_bins = read(input, Int32)
        local pbin::PseudoBin
        for i in 1:n_bins
            bin = read(input, UInt32)
            n_chunks = read(input, Int32)
            if bin == 37450
                # Pseudo-bin:
                # This is not mentioned in tabix's specs but exists in real data.
                @assert n_chunks == 2
                cnk_beg = read(input, UInt64)
                cnk_end = read(input, UInt64)
                n_mapped = read(input, UInt64)
                n_unmapped = read(input, UInt64)
                pbin = PseudoBin(Chunk(cnk_beg, cnk_end), n_mapped, n_unmapped)
            else
                chunks = Chunk[]
                for j in 1:n_chunks
                    cnk_beg = read(input, UInt64)
                    cnk_end = read(input, UInt64)
                    push!(chunks, Chunk(cnk_beg, cnk_end))
                end
                binindex[bin] = chunks
            end
        end

        n_intvs = read(input, Int32)
        linindex = LinearIndex()
        for i in 1:n_intvs
            ioff = read(input, UInt64)
            push!(linindex, ioff)
        end

        push!(indexes, (binindex, linindex, pbin))
    end

    if !eof(input)
        n_no_coor = Nullable(read(input, UInt64))
    else
        n_no_coor = Nullable{UInt64}()
    end

    return Tabix(
        format,
        (col_seq, col_beg, col_end),
        meta,
        skip,
        names,
        indexes,
        n_no_coor)
end

function Base.seek(stream::BGZFStream, chunk::Chunk)
    seek(stream, chunk.start)
end

# The BED rule is half-closed-half-open and 0-based like Python.
function is_bed_rule(format)
    return format & 0x10000 != 0
end

function format2str(format)
    if format == 1
        return "SAM"
    elseif format == 2
        return "VCF"
    else
        if is_bed_rule(format)
            return "generic (BED rule)"
        else
            return "generic"
        end
    end
end

const LinearWindowSize = 16 * 1024

"""
    overlapchunks(index::Tabix, seqname::String, interval::UnitRange)
    overlapchunks(index::Tabix, seqid::Integer, interval::UnitRange)

Return chunks possibly overlapping the range specified by `seqname` (or `seqid`)
and `interval`; `seqid` and `interval` must be 1-based index and inclusive.

NOTE: Records within the returned chunks are not guaranteed to overlap the query
region. The user need to check the condition as:
    line = readline(stream)
    values = split(chomp(line), '\t')
    seqname = value[index.columns[1]]
    start = parse(Int, index.columns[2])
    stop = parse(Int, index.columns[3]) - 1  # needs -1 for the BED format
    if seqname == target_seqname && intersect(start:stop, target_interval)
        # the line overlaps the query region
    end
"""
function overlapchunks(index::Tabix, seqid::Integer, interval::UnitRange)
    if !(1 ≤ seqid ≤ endof(index.indexes))
        throw(ArgumentError("sequence id $(seqid) is out of range"))
    end

    if isempty(interval)
        return Chunk[]
    end

    binindex, linindex = index.indexes[seqid]
    bins = reg2bins(first(interval), last(interval))
    offset = linindex[cld(first(interval), LinearWindowSize)]
    ret = Chunk[]
    for bin in bins
        if haskey(binindex, bin)
            for chunk in binindex[bin]
                if chunk.stop > offset
                    push!(ret, chunk)
                end
            end
        end
    end

    sort!(ret)
    reduce!(ret)

    return ret
end

function overlapchunks(index::Tabix, seqname::AbstractString, interval::UnitRange)
    seqid = findfirst(index.names, seqname)
    if seqid == 0
        throw(ArgumentError("sequence name $(seqname) is not included in the index"))
    end
    return overlapchunks(index, seqid, interval)
end

# Merge chunks to minimize the number of seek operations.
function reduce!(chunks)
    @assert issorted(chunks)
    i = 1
    while i < endof(chunks)
        chunk = chunks[i]
        next = chunks[i+1]
        if chunk.stop < next.start
            # neither overlapping nor adjacent
            i += 1
            continue
        end
        chunks[i] = Chunk(chunk.start, max(chunk.stop, next.stop))
        deleteat!(chunks, i + 1)
    end
    return chunks
end

# Calculate bins overlapping a region [from, to] (one-based).
function reg2bins(from, to)
    bins = UInt32[]
    bin_start = 0
    for scale in 29:-3:14
        for k in ((from - 1) >> scale):((to - 1) >> scale)
            push!(bins, bin_start + k)
        end
        bin_start = 8 * bin_start + 1
    end
    return bins
end


src/bgzfstream.jl
# BGZFStream
# ==========

# Internal details
# ----------------
#
# When reading data from an input, compressed data will be read to a buffer
# (compressed block) and then inflated into a decompressed block at a time.
# When writing data to an output, raw data will be deflated into a compressed
# block and then written to the output immediately.  Each data block is no
# larger than 64 KiB before and after compression.
#
# Read mode (stream.mode = READ_MODE)
# -----------------------------------
#
#          compressed block          decompressed block
# stream   +---------------+         +---------------+
# .io ---> |xxxxxxx        | ------> |xxxxxxxxxxx    | --->
#     read +---------------+ inflate +---------------+ read
#                                    |------>| block.position ∈ [0, 64K)
#                                    |--------->| block.size ∈ [0, 64K]
#
# Write mode (stream.mode = WRITE_MODE)
# -------------------------------------
#
#          compressed block          decompressed block
# stream   +---------------+         +---------------+
# .io <--- |xxxxxxx        | <------ |xxxxxxxx       | <---
#    write +---------------+ deflate +---------------+ write
#                                    |------>| block.position ∈ [0, 64K)
#                                    |------------->| block.size = 64K - 256
# - xxx: used data
# - 64K: 65536 (= BGZF_MAX_BLOCK_SIZE = 64 * 1024)

mutable struct Block
    # space for the compressed block
    compressed_block::Vector{UInt8}

    # space for the decompressed block
    decompressed_block::Vector{UInt8}

    # block offset in a file (this is always 0 for a pipe stream)
    block_offset::Int

    # the next reading byte position in a block
    position::Int

    # number of available bytes in the decompressed block
    size::Int

    # zstream object
    zstream::Libz.ZStream
end

function Block(mode)
    compressed_block = Vector{UInt8}(BGZF_MAX_BLOCK_SIZE)
    decompressed_block = Vector{UInt8}(BGZF_MAX_BLOCK_SIZE)

    if mode == READ_MODE
        zstream = Libz.init_inflate_zstream(true)
        size = 0
    else
        zstream = Libz.init_deflate_zstream(
            true,
            Libz.Z_DEFAULT_COMPRESSION,
            8,  # default memory level
            Libz.Z_DEFAULT_STRATEGY)
        size = BGZF_SAFE_BLOCK_SIZE
    end

    return Block(compressed_block, decompressed_block, 0, 1, size, zstream)
end

# Stream type for the BGZF compression format.
mutable struct BGZFStream{T<:IO} <: IO
    # underlying IO stream
    io::T

    # read/write mode
    mode::UInt8

    # compressed & decompressed blocks with metadata and zstream
    blocks::Vector{Block}

    # current block index
    block_index::Int

    # whether stream is open
    isopen::Bool

    # callback function called when closing the stream
    onclose::Function
end

# BGZF blocks are no larger than 64 KiB before and after compression.
const BGZF_MAX_BLOCK_SIZE = UInt(64 * 1024)

# BGZF_MAX_BLOCK_SIZE minus "margin for safety"
# NOTE: Data block will become slightly larger after deflation when bytes are
# randomly distributed.
const BGZF_SAFE_BLOCK_SIZE = UInt(BGZF_MAX_BLOCK_SIZE - 256)

# Read mode:  inflate and read a BGZF file
# Write mode: deflate and write a BGZF file
const READ_MODE  = 0x00
const WRITE_MODE = 0x01

"""
    BGZFStream(io::IO[, mode::AbstractString="r"])
    BGZFStream(filename::AbstractString[, mode::AbstractString="r"])

Create an I/O stream for the BGZF compression format.

The first argument is either an `IO` object or a filename. If `mode` is `"r"`
(read) the BGZF stream will be in read mode and decompress the underlying BGZF
blocks while reading. In read mode, `BGZFStream` supports the `seek` operation
using a virtual file offset (see `VirtualOffset`). If `mode` is `"w"` (write)
or `"a"` (append) the BGZF stream will be in write mode and compress written
data to BGZF blocks.
"""
function BGZFStream(io::IO, mode::AbstractString="r")
    if mode ∉ ("r", "w", "a")
        throw(ArgumentError("invalid mode: \"", mode, "\""))
    end

    # the number of parallel workers
    mode′ = mode == "r" ? READ_MODE : WRITE_MODE
    if mode′ == READ_MODE
        blocks = [Block(mode′) for _ in 1:Threads.nthreads()]
    else
        # Write mode is not (yet?) multi-threaded.
        blocks = [Block(mode′)]
    end
    return BGZFStream(io, mode′, blocks, 1, true, io -> close(io))
end

function BGZFStream(filename::AbstractString, mode::AbstractString = "r")
    if mode ∉ ("r", "w", "a")
        throw(ArgumentError("invalid mode: '", mode, "'"))
    end
    return BGZFStream(open(filename, mode), mode)
end

function Base.open(::Type{BGZFStream}, filepath::AbstractString, mode::AbstractString = "r")
    return BGZFStream(filepath, mode)
end

"""
    virtualoffset(stream::BGZFStream)

Return the current virtual file offset of `stream`.
"""
function virtualoffset(stream::BGZFStream)
    if stream.mode == READ_MODE
        i = ensure_buffered_data(stream)
        if i == 0
            block = stream.blocks[stream.block_index]
        else
            block = stream.blocks[i]
        end
    else
        block = stream.blocks[1]
    end
    return VirtualOffset(block.block_offset, block.position - 1)
end

function virtualoffset(stream::BGZFStream{Pipe})
    throw(ArgumentError("virtualoffset is not supported for a pipe stream"))
end

function Base.show(io::IO, stream::BGZFStream)
    print(io,
        summary(stream),
        "(<",
        "mode=", stream.mode == READ_MODE ? "read" : "write",
        ">)")
end

function Base.isopen(stream::BGZFStream)
    return stream.isopen
end

function Base.close(stream::BGZFStream)
    if stream.mode == WRITE_MODE
        if stream.blocks[1].position > 1
            write_blocks!(stream)
        end
        write(stream.io, EOF_BLOCK)
    end
    for block in stream.blocks
        end_zstream(block.zstream, stream.mode)
    end
    stream.isopen = false
    stream.onclose(stream.io)
    return
end

function Base.flush(stream::BGZFStream)
    if stream.mode == WRITE_MODE
        flush(stream.io)
    end
    return
end

function Base.eof(stream::BGZFStream)
    if stream.mode == READ_MODE
        return ensure_buffered_data(stream) == 0
    else
        return true
    end
end

function Base.seekstart(stream::BGZFStream)
    seek(stream, VirtualOffset(0, 0))
end

function Base.seek(stream::BGZFStream, voffset::VirtualOffset)
    if stream.mode == WRITE_MODE
        throw(ArgumentError("BGZFStream in write mode is not seekable"))
    end
    block_offset, inblock_offset = offsets(voffset)
    seek(stream.io, block_offset)
    read_blocks!(stream)
    block = first(stream.blocks)
    if inblock_offset ≥ block.size
        throw(ArgumentError("too large in-block offset"))
    end
    block.block_offset = block_offset
    block.position = inblock_offset + 1
    return
end

function Base.seek(stream::BGZFStream{Pipe}, voffset::VirtualOffset)
    throw(ArgumentError("seek is not supported for a pipe stream"))
end

function Base.read(stream::BGZFStream, ::Type{UInt8})
    if !isopen(stream)
        throw(ArgumentError("stream is already closed"))
    elseif stream.mode != READ_MODE
        throw(ArgumentError("stream is not readable"))
    end
    block_index = ensure_buffered_data(stream)
    if block_index == 0
        throw(EOFError())
    end
    block = stream.blocks[block_index]
    byte = block.decompressed_block[block.position]
    block.position += 1
    return byte
end

function Base.write(stream::BGZFStream, byte::UInt8)
    if !isopen(stream)
        throw(ArgumentError("stream is already closed"))
    elseif stream.mode != WRITE_MODE
        throw(ArgumentError("stream is not writable"))
    end
    block = stream.blocks[1]
    block.decompressed_block[block.position] = byte
    block.position += 1
    if block.position > block.size
        write_blocks!(stream)
    end
    return 1
end

function Base.unsafe_read(stream::BGZFStream, p::Ptr{UInt8}, n::UInt)
    if !isopen(stream)
        throw(ArgumentError("stream is already closed"))
    elseif stream.mode != READ_MODE
        throw(ArgumentError("stream is not readable"))
    end
    p_end = p + n
    while p < p_end
        i = ensure_buffered_data(stream)
        if i == 0
            throw(EOFError())
        end
        @inbounds block = stream.blocks[i]
        len = min(p_end - p, block.size - block.position + 1)
        src = pointer(block.decompressed_block, block.position)
        memcpy(p, src, len)
        block.position += len
        p += len
    end
end

function Base.unsafe_write(stream::BGZFStream, p::Ptr{UInt8}, n::UInt)
    if !isopen(stream)
        throw(ArgumentError("stream is already closed"))
    elseif stream.mode != WRITE_MODE
        throw(ArgumentError("stream is not writable"))
    end
    block = stream.blocks[1]
    p_end = p + n
    while p < p_end
        len = min(p_end - p, block.size - block.position + 1)
        dst = pointer(block.decompressed_block, block.position)
        memcpy(dst, p, len)
        block.position += len
        if block.position > block.size
            write_blocks!(stream)
        end
        p += len
    end
    return Int(n)
end


# Internal functions
# ------------------

# Ensure buffered data (at least 1 byte) for reading and return the block index
# if available or 0 otherwise.
@inline function ensure_buffered_data(stream)
    #@assert stream.mode == READ_MODE
    @label doit
    while stream.block_index ≤ endof(stream.blocks)
        @inbounds block = stream.blocks[stream.block_index]
        if block.position ≤ block.size
            return stream.block_index
        end
        stream.block_index += 1
    end
    if !eof(stream.io)
        read_blocks!(stream)
        @goto doit
    end
    return 0
end

# A wrapper of memcpy.
function memcpy(dst, src, len)
    ccall(
        :memcpy,
        Ptr{Void},
        (Ptr{Void}, Ptr{Void}, Csize_t),
        dst, src, len)
end

struct BGZFDataError <: Exception
    message::AbstractString
end

# Throw a BGZFDataError exception with the given error message.
function bgzferror(message::AbstractString="malformed BGZF data")
    throw(BGZFDataError(message))
end

# Read and inflate blocks.
function read_blocks!(stream)
    @assert stream.mode == READ_MODE

    # read BGZF blocks in sequence
    n_blocks = 0
    has_position = true
    try
        position(stream.io)
    catch
        has_position = false
    end
    while n_blocks < length(stream.blocks) && !eof(stream.io)
        block = stream.blocks[n_blocks += 1]
        if has_position
            block.block_offset = position(stream.io)
        end
        block.position = 1
        bsize = read_bgzf_block!(stream.io, block.compressed_block)
        zstream = block.zstream
        zstream.next_in = pointer(block.compressed_block)
        zstream.avail_in = bsize
        zstream.next_out = pointer(block.decompressed_block)
        zstream.avail_out = BGZF_MAX_BLOCK_SIZE
    end

    # inflate blocks in parallel
    rets = Vector{Cint}(n_blocks)
    Threads.@threads for i in 1:n_blocks
        block = stream.blocks[i]
        zstream = block.zstream
        old_avail_out = zstream.avail_out
        rets[i] = Libz.inflate!(zstream, Libz.Z_FINISH)
        block.size = old_avail_out - zstream.avail_out
    end

    for i in 1:n_blocks
        if rets[i] != Libz.Z_STREAM_END
            error("zlib failed to inflate a compressed block")
        end
        block = stream.blocks[i]
        @assert block.size ≤ BGZF_MAX_BLOCK_SIZE
        reset_zstream(block.zstream, stream.mode)
    end

    stream.block_index = 1
    return
end

# Read a BGZF block from `input`.
function read_bgzf_block!(input, block)
    # TODO: check the number of read bytes

    # +---+---+---+---+---+---+---+---+---+---+
    # |ID1|ID2|CM |FLG|     MTIME     |XFL|OS | (more-->)
    # +---+---+---+---+---+---+---+---+---+---+
    unsafe_read(input, pointer(block), 10)
    id1_ok = block[1] == 0x1f
    id2_ok = block[2] == 0x8b
    cm_ok  = block[3] == 0x08
    flg_ok = block[4] == 0x04
    if !id1_ok || !id2_ok
        bgzferror("invalid gzip identifier")
    elseif !cm_ok
        bgzferror("invalid compression method")
    elseif !flg_ok
        bgzferror("invalid flag")
    end

    # +---+---+=================================+
    # | XLEN  |...XLEN bytes of "extra field"...| (more-->)
    # +---+---+=================================+
    unsafe_read(input, pointer(block, 11) , 2)
    xlen = UInt16(block[11]) | UInt16(block[12]) << 8
    unsafe_read(input, pointer(block, 13), xlen)
    bsize::Int = 0
    pos = 12
    while pos < 12 + xlen
        si1 = block[pos+1]
        si2 = block[pos+2]
        slen = UInt16(block[pos+3]) | UInt16(block[pos+4]) << 8
        if si1 == 0x42 || si2 == 0x43
            if slen != 2
                bgzferror("invalid subfield length")
            end
            bsize = (UInt16(block[pos+5]) | UInt16(block[pos+6]) << 8) + 1
        end
        # skip this field
        pos += 4 + slen
    end
    if bsize == 0
        bgzferror("no block size")
    end

    # +=======================+---+---+---+---+---+---+---+---+
    # |...compressed blocks...|     CRC32     |     ISIZE     |
    # +=======================+---+---+---+---+---+---+---+---+
    size = bsize - 1 - xlen - 19 + 8
    unsafe_read(input, pointer(block, 13 + xlen), size)

    if eof(input) && !is_eof_block(block)
        bgzferror("no end-of-file marker (maybe a truncated file)")
    end

    return bsize
end

function write_blocks!(stream)
    @assert stream.mode == WRITE_MODE

    n_blocks = length(stream.blocks)
    @assert n_blocks == 1

    for i in 1:n_blocks
        block = stream.blocks[i]
        zstream = block.zstream
        zstream.next_in = pointer(block.decompressed_block)
        zstream.avail_in = block.position - 1
        zstream.next_out = pointer(block.compressed_block, 9)
        zstream.avail_out = BGZF_MAX_BLOCK_SIZE - 8

        ret = Libz.deflate!(zstream, Libz.Z_FINISH)
        if ret != Libz.Z_STREAM_END
            if ret == Libz.Z_OK
                error("block size may exceed BGZF_MAX_BLOCK_SIZE")
            else
                error("failed to compress a BGZF block (zlib error $(ret))")
            end
        end

        blocksize = (BGZF_MAX_BLOCK_SIZE - 8) - zstream.avail_out + 8
        fix_header!(block.compressed_block, blocksize)
        nb = unsafe_write(stream.io, pointer(block.compressed_block), blocksize)
        if nb != blocksize
            error("failed to write a BGZF block")
        end
        if !isa(stream.io, Pipe)
            block.block_offset = position(stream.io)
        end
        block.position = 1

        reset_zstream(zstream, WRITE_MODE)
    end
end

function fix_header!(block, blocksize)
    copy!(block,
          # ID1   ID2    CM   FLG  |<--     MTIME    -->|   XFL    OS
          [0x1f, 0x8b, 0x08, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff])
    copy!(block, 11,
          #  XLEN    S1    S2    SLEN          BSIZE
          reinterpret(UInt8, [0x0006, 0x4342, 0x0002, UInt16(blocksize - 1)]))
end

# end-of-file marker block (used for detecting unintended file truncation)
const EOF_BLOCK = [
    0x1f, 0x8b, 0x08, 0x04, 0x00, 0x00, 0x00, 0x00,
    0x00, 0xff, 0x06, 0x00, 0x42, 0x43, 0x02, 0x00,
    0x1b, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00
]

# Return true iff the block is a end-of-file marker.
function is_eof_block(block)
    if length(block) < length(EOF_BLOCK)
        return false
    end
    for i in 1:endof(EOF_BLOCK)
        if block[i] != EOF_BLOCK[i]
            return false
        end
    end
    return true
end

# Reset the zstream.
function reset_zstream(zstream, mode)
    if mode == READ_MODE
        Libz.@zcheck Libz.reset_inflate!(zstream)
    else
        Libz.@zcheck Libz.reset_deflate!(zstream)
    end
    return
end

# End the zstream.
function end_zstream(zstream, mode)
    if mode == READ_MODE
        Libz.@zcheck Libz.end_inflate!(zstream)
    else
        Libz.@zcheck Libz.end_deflate!(zstream)
    end
    return
end
